src/web/core/src/main/java/org/geoserver/web/WicketConfigurationLockCallback.java

```diff
@@ -45,30 +47,59 @@ public void onEndRequest() {
         LockType type = THREAD_LOCK.get();
         if (type != null) {
             THREAD_LOCK.remove();
+
+            if (THREAD_REQUESTED_PAGE.get() != null) {
+                THREAD_REQUESTED_PAGE.remove();
+            }
+            
             locker.unlock(type);
         }
     }
 
     @Override
     public void onRequestTargetSet(Class<? extends IRequestablePage> requestTarget) {
-        // we can have many of these calls per http call, avoid locking multiple times,
-        // onEndRequest will be called just once
-        LockType type = THREAD_LOCK.get();
-        if (type != null || requestTarget == null) {
-            return;
+        if (!GeoServerUnlockablePage.class.isAssignableFrom(requestTarget)) {
+            // we can have many of these calls per http call, avoid locking multiple times,
+            // onEndRequest will be called just once
+            LockType type = THREAD_LOCK.get();
+            if (type != null || requestTarget == null) {
+                return;
+            }
+    
+            // setup a write lock for secured pages, a read one for the others
+            if (GeoServerSecuredPage.class.isAssignableFrom(requestTarget)) {
+                type = LockType.WRITE;
+            }
+            if (type == null) {
+                type = LockType.READ;
+            }
+    
+            // and lock
+            boolean lockTaken = locker.tryLock(type);
+    
+            if (lockTaken) {
+                THREAD_LOCK.set(type);
+                THREAD_REQUESTED_PAGE.set(requestTarget);
+            }
         }
+    }
+    
+    @Override
+    public void onRequestTargetSet(RequestCycle cycle,
+            Class<? extends IRequestablePage> requestTarget) {
+        WicketCallback.super.onRequestTargetSet(cycle, requestTarget);
```

Comment:
I'd call directly the other method, or just merge them into one, after all, the other method is not going to get called anymore from GeoServerApplication right?